File: aulas/aula01_introducao/lab3_ecommerce.py
class Produto:
    def __init__(self, nome, preco):
        self.nome = nome
        self.preco = preco
        
    def aplicar_desconto(self, porcentagem):
        valorDesconto = (porcentagem / 100) * self.preco
        self.preco = self.preco - valorDesconto
        
        print(f"O desconto foi de {porcentagem}%")
        print(f"O valor do produto agora é: R${self.preco} \n")
        
    def exibir_dados(self):
        print(f"Nome do produto {self.nome}")
        print(f"Preco do produto: {self.preco}")
        
class Livro(Produto):
    def __init__(self, nome, preco, autor):
        super().__init__(nome, preco)
        self.autor = autor
        
    def exibir_dados(self):
        super().exibir_dados()
        print(f"Autor {self.autor} \n")
        
class Eletronico(Produto):
    def __init__(self, nome, preco, voltagem):
        super().__init__(nome, preco)
        self.voltagem = voltagem
        
    def exibir_dados(self):
        super().exibir_dados()
        print(f"Voltagem {self.voltagem} \n")
        
l1 = Livro("O Destino Chega", 49.99, "MARVEL")
e1 = Eletronico("Arduino UNO Q", 359.9, 3.3)

l1.exibir_dados()
l1.aplicar_desconto(30)
l1.exibir_dados()

e1.exibir_dados()
e1.aplicar_desconto(45)
e1.exibir_dados()

        

File: aulas/aula01_introducao/C++/lab3_ecommerce.cpp
#include <iostream>
#include <string>
#include <memory>

class Produto {
private:
    double desconto;

protected:
    std::string nome;
    double preco;

public:
    Produto (std::string n, double p)
        : nome(n), preco(p) {}

    virtual ~Produto() = default;

    void aplicar_desconto(double porcentagem) {
        desconto = (porcentagem / 100) * preco;
        preco = preco - desconto;

        std::cout << "Desconto no produto: " << nome << "\n"
                  << "O desconto foi de: " << porcentagem << "%" << "\n"
                  << "O valor do produto agora e: R$" << preco << "\n\n"; 

    }

    virtual void exibir_dados() {
        std::cout << "Nome do produto: " << nome << "\n"
                  << "Preco do produto: R$" << preco << "\n\n";   
    }
};

class Livro : public Produto {
private:
std::string autor;

public:
    Livro (std::string n, double p, std::string a) 
        : Produto(n, p), autor(a) {}

    void exibir_dados() override {
        std::cout << "Nome do produto: " << nome << "\n"
                  << "Preco do produto: R$" << preco << "\n"
                  << "Autor: " << autor << "\n\n";
    }
};

class Eletronico : public Produto {
private:
    double voltagem;

public:
    Eletronico (std::string n, double p, double v)
        : Produto(n, p), voltagem(v) {}

    void exibir_dados() override {
        std::cout << "Nome do produto: " << nome << "\n"
                  << "Preco do produto: R$" << preco << "\n"
                  << "Voltagem: " << voltagem << "\n\n";
    }
};

int main() {
    std::unique_ptr<Produto> p1 = std::make_unique<Livro>("O Destino Chega", 49.99, "MARVEL");
    std::unique_ptr<Produto> p2 = std::make_unique<Eletronico>("Arduino UNO Q", 359.9, 3.3);

    p1->exibir_dados();
    p1->aplicar_desconto(30);
    p1->exibir_dados();

    p2->exibir_dados();
    p2->aplicar_desconto(45);
    p2->exibir_dados();
    return 0;
}